jetty-io/src/main/java/org/eclipse/jetty/io/AbstractByteBufferPool.java

```diff
@@ -30,36 +35,64 @@
 abstract class AbstractByteBufferPool implements ByteBufferPool
 {
     private final int _factor;
-    private final int _maxQueueLength;
+    private final int _maxCapacity;
+    private final int _maxBucketSize;
     private final long _maxHeapMemory;
     private final long _maxDirectMemory;
     private final AtomicLong _heapMemory = new AtomicLong();
     private final AtomicLong _directMemory = new AtomicLong();
-
+    private final RetainableByteBufferPool _retainableByteBufferPool;
+    
     /**
      * Creates a new ByteBufferPool with the given configuration.
      *
      * @param factor the capacity factor
-     * @param maxQueueLength the maximum ByteBuffer queue length
+     * @param maxBucketSize the maximum ByteBuffer queue length
      * @param maxHeapMemory the max heap memory in bytes, -1 for unlimited memory or 0 to use default heuristic
      * @param maxDirectMemory the max direct memory in bytes, -1 for unlimited memory or 0 to use default heuristic
+     * @param retainedHeapMemory the max heap memory in bytes, -1 for no retained memory or 0 to use default heuristic
+     * @param retainedDirectMemory the max direct memory in bytes, -1 for no retained memory or 0 to use default heuristic
      */
-    protected AbstractByteBufferPool(int factor, int maxQueueLength, long maxHeapMemory, long maxDirectMemory)
+    protected AbstractByteBufferPool(int factor, int maxCapacity, int maxBucketSize, long maxHeapMemory, long maxDirectMemory, long retainedHeapMemory, long retainedDirectMemory)
     {
         _factor = factor <= 0 ? 1024 : factor;
-        _maxQueueLength = maxQueueLength;
+        _maxCapacity = maxCapacity > 0 ? maxCapacity : 64 * _factor;
+        _maxBucketSize = maxBucketSize;
         _maxHeapMemory = (maxHeapMemory != 0) ? maxHeapMemory : Runtime.getRuntime().maxMemory() / 4;
         _maxDirectMemory = (maxDirectMemory != 0) ? maxDirectMemory : Runtime.getRuntime().maxMemory() / 4;
+
+        if (retainedHeapMemory < 0 && retainedDirectMemory < 0)
```

Comment:
I don't think so.  The main byte buffer pool is keep hold of idle buffers only - ie ones that it has previously been acquired and released.   The retainableByteBufferPool is kind of just like a client that has acquired but not yet released the buffer.

So clear should not release retained buffers any more than it should "take back" buffers allocated to other users.    But perhaps it could trigger a shrinking of the retained pool to only buffers that are in use?   Is t hat semantic already supported?

Reply:
@lorban I think we need to review exactly what clear does on a retainable pool.... which means we need to review what remove does on a Pool.  I don't think the semantic for that is correct, specially for removing active items.  Also for multi items, remove might return false, but the item is actually closed so will eventually be removed?? this makes accounting on the return value of remove difficult.   Needs review.   So for now, let's not wire up clear.

Reply:
The `clear` contracts of both pools are slightly different, but not incompatible IMHO.

`clear` on BBP immediately stops holding onto its buffers.
`clear` on RBBP immediately stops holding onto the unused buffers it contains, and marks the ones still in use as to-be-removed-when-released.

If you clear the RBBP then clear the BBP you make RBBP give all its unused buffers back to the BBP then make BBP drop all its buffers. I think this is what we want.